### PR TITLE
Lead and Lag window functions should support default value with datatype other than Int64

### LINKED ISSUE #8307: `lag` and `lead` window functions does not accept compatible default value
### Describe the bug

The window functions `lag` and `lead` allow a default value as its third parameter.
Currently in DataFusion, this needs to be Int64 even though the value selected in the function is not of type Int64.

Postgres docs say that the default value only has to be compatible: https://www.postgresql.org/docs/current/functions-window.html

### To Reproduce

```
DataFusion CLI v33.0.0
❯ create table t1 (a timestamp);
0 rows in set. Query took 0.009 seconds.

❯ insert into t1 values (1),(2),(3);
+-------+
| count |
+-------+
| 3     |
+-------+
1 row in set. Query took 0.013 seconds.

❯ select lag(a, 1, '1970-01-01T00:00:00.000000001'::timestamp) OVER(ORDER BY a) from t1;
Internal error: Expects default value to have Int64 type.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### Expected behavior

Allow any default value that can be coerced into the first parameter

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #8307.

## Rationale for this change

Noticed this when reviewing #8989. Currently we only support default value with Int64 data type for Lead and Lag window functions. This patch enables other data types to be used as default value.

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -2364,6 +2364,16 @@ impl ScalarValue {
         ScalarValue::try_from_array(&cast_arr, 0)
     }
 
+    /// Try to cast this value to a ScalarValue of type `data_type`
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:2367]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me -- thanks @viirya 

cc @mustafasrepo and @comphead

**Review by mustafasrepo — APPROVED** [review]

**Comment by simonvandel** [comment]
Thanks! I think this is the fix to https://github.com/apache/arrow-datafusion/issues/8307, so you can use that issue number in the PR description.

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -2364,6 +2364,16 @@ impl ScalarValue {
...
+    pub fn cast_to(&self, data_type: &DataType) -> Result<Self> {
```
Comment (comphead) [inline datafusion/common/src/scalar.rs:2368]: I was doing the same to overcome unnecessary string conversion 👍

**Review by comphead — APPROVED** [review]
lgtm

**Comment by viirya** [comment]
Thank you @alamb @mustafasrepo @simonvandel @comphead

**Pull request merged**